### PR TITLE
Feature/say formatted

### PR DESCRIPTION
**Pull request by jlunenburg**

### CODE REVIEW

**Comment by jlunenburg** [comment]
@MatthijsBurgh I added a commit in which Say inherits from SayFormatted to remove duplicated code and added @LarsJanssenTUe as reviewer

**Comment by LoyVanBeek** [comment]
Are there arguments against not renaming `SayFormatted` to just `Say`? I was opposed to this because of the added complexity of the `-Formatted` version, but maintaining 2 versions is also not nice. This PR solves that. 
The added complexity is also low and optional, so maybe just do the rename?

**Comment by MatthijsBurgh** [comment]
Looks good to me, but indeed @LarsJanssenTUe should review this thoroughly.

**Comment by jlunenburg** [comment]
> Are there arguments against not renaming `SayFormatted` to just `Say`? I was opposed to this because of the added complexity of the `-Formatted` version, but maintaining 2 versions is also not nice. This PR solves that.
> The added complexity is also low and optional, so maybe just do the rename?

Renaming would also work imho. You even could add (for backwards compatibility) `SayFormatted = Say`

**Comment by LoyVanBeek** [comment]
Could work, but `SayFormatted` still has only a few usages while `Say` has a lot. 

I'd say: go for the rename of `SayFormatted` to just `Say` and replace usages of `SayFormatted` with `Say`

File: robot_smach_states/src/robot_smach_states/human_interaction/human_interaction.py [hunk 1]
```diff
@@ -65,21 +82,75 @@ def execute(self, userdata=None):
...
+    >>>
+    >>> say = Say(robot, ["a", "b", "c"])
+    >>> #Repeat command 50 times, every time it should succeed and return "spoken"
+    >>> outcomes = [say.execute() for i in range(50)]
+    >>> assert all(outcome == "spoken" for outcome in outcomes)
+    >>>
+    >>> say2 = Say(robot, ds.VariableDesignator('aap'))
+    >>> say2.execute()
+    'spoken'
+    >>> robot.speech.speak.assert_called_with('aap', None, None, None, None, True)
```
Comment (LarsJanssenTUe) [inline robot_smach_states/src/robot_smach_states/human_interaction/human_interaction.py]: The assert_called_with does not exist as far as I know

File: robot_smach_states/src/robot_smach_states/human_interaction/human_interaction.py [hunk 2]
```diff
@@ -65,21 +82,75 @@ def execute(self, userdata=None):
...
+    >>>
+    >>> d1 = ds.VariableDesignator(resolve_type=str).writeable
+    >>> d1.write('banana')
+    >>> say3 = Say(robot, d1)
+    >>> say3.execute()
+    'spoken'
+    >>> robot.speech.speak.assert_called_with('banana', None, None, None, None, True)
```
Comment (LarsJanssenTUe) [inline robot_smach_states/src/robot_smach_states/human_interaction/human_interaction.py]: see previous comment

**Comment by LarsJanssenTUe** [comment]
A seperate test file is missing, so that should be added

File: robot_smach_states/src/robot_smach_states/human_interaction/human_interaction.py (hunk 1 above)
Comment (MatthijsBurgh) [reply robot_smach_states/src/robot_smach_states/human_interaction/human_interaction.py]: This is part of MagicMock

File: robot_smach_states/src/robot_smach_states/human_interaction/human_interaction.py (hunk 2 above)
Comment (MatthijsBurgh) [reply robot_smach_states/src/robot_smach_states/human_interaction/human_interaction.py]: Same, part of MagicMock

File: robot_smach_states/src/robot_smach_states/human_interaction/human_interaction.py [hunk 3]
```diff
@@ -65,21 +82,75 @@ def execute(self, userdata=None):
             rospy.logerr("sentence = None, not saying anything...")
             return "spoken"
 
-        if not isinstance(self.sentence, str) and isinstance(self.sentence, list):
-            self.sentence = random.choice(self.sentence)
+        if hasattr(self.sentence, "resolve"):
```
[the remark this answers is not in the document]
Comment (MatthijsBurgh) [reply robot_smach_states/src/robot_smach_states/human_interaction/human_interaction.py]: Of dat is gewoon prima, error melding is goed. Dus laat maar stuk gaan.

File: robot_smach_states/src/robot_smach_states/human_interaction/human_interaction.py [hunk 4]
```diff
@@ -15,39 +16,55 @@
...
-        # ds.check_type(mood, str)
-        ds.check_type(block, bool)
+        ds.check_type(sentence, [str], str)
+        assert(isinstance(language, str) or isinstance(language, type(None)))
+        assert(isinstance(personality, str) or isinstance(personality, type(None)))
+        assert(isinstance(voice, str) or isinstance(voice, type(None)))
+        assert(isinstance(mood, str) or isinstance(mood, type(None)))
+        assert(isinstance(block, bool))
+
+        assert(all(isinstance(v, ds.Designator) for v in place_holders.values()))
```
[the remark this answers is not in the document]
Comment (MatthijsBurgh) [reply robot_smach_states/src/robot_smach_states/human_interaction/human_interaction.py]: Done

**Comment by MatthijsBurgh** [comment]
I would like to discuss the testing with somebody else. As we will be creating more tests, we will need a structure for it.

**Comment by MatthijsBurgh** [comment]
Also the designator check is broken in master, it is fixed in the rwc2019 branch of robot_smach_states.

File: robot_smach_states/src/robot_smach_states/util/designators/checks.py [hunk 5]
```diff
@@ -59,12 +86,39 @@ def check_type(designator_or_value, *allowed_types):
...
+        if isinstance(designator_or_value, list):
+            allowed_list_types = [allowed_type[0] for allowed_type in allowed_types if isinstance(allowed_type, list)]
+
+            if type(designator_or_value[0]) not in allowed_list_types:
+                allowed_list_types_string = ", ".join(map("[{}]".format, allowed_list_types))
+                raise TypeError("{0} is of type [{1}] but should be {2}".format(designator_or_value,
+                                                                                type(designator_or_value[0]),
+                                                                                allowed_list_types_string))
+        else:
+            if not type(designator_or_value) in allowed_types:
```
Comment (LarsJanssenTUe) [inline robot_smach_states/src/robot_smach_states/util/designators/checks.py]: Shouldn't allowed_types in this if also be the the allowed_list_types?

Comment (MatthijsBurgh) [reply robot_smach_states/src/robot_smach_states/util/designators/checks.py]: No, because the value isn't a list here.

**Pull request merged**